design/unified-repo-and-kopia-integration/unified-repo-and-kopia-integration.md

```diff
@@ -0,0 +1,425 @@
+    ///Delete a manifest object
+    DeleteManifest(id ID) error
+ 
+    ///Flush all the backup repository data
+    Flush() error
+ 
+    ///Get the local time of the backup repository. It may be different from the time of the caller
+    Time() time.Time
+ 
+    ///Close the backup repository
+    Close() error
+}
+
+type ObjectReader interface {
+    io.ReadCloser
+    io.Seeker
+ 
+    ///Length returns the logical size of the object
+    Length() int64
+}
+ 
+type ObjectWriter interface {
+    io.WriteCloser
+ 
+    ///For some cases, i.e. block incremental, the object is not written sequentially
+    io.Seeker
+ 
+ 
+    ///Wait for the completion of the object write
+    ///Result returns the object's unified identifier after the write completes
+    Result() (ID, error)
+}  
+```
+ 
+Some data structure & constants used by the interfaces:  
+```  
+type RepoOptions struct {
+   ///A repository specific string to identify a backup storage, i.e., "s3", "filesystem"
+   StorageType string
+   ///Backup repository password, if any
+   RepoPassword string
+   ///A custom path to save the repository's configuration, if any
+   ConfigFilePath string
+   ///Other repository specific options
+   GeneralOptions map[string]string
+   ///Storage specific options
+   StorageOptions map[string]string
+}
+
+///ObjectWriteOptions defines the options when creating an object for write
+type ObjectWriteOptions struct {
+    FullPath    string ///Full logical path of the object
+    Description string ///A description of the object, could be empty
+    Prefix      ID     ///A prefix of the name used to save the object
+    AccessMode  int    ///OBJECT_DATA_ACCESS_*
+    BackupMode  int    ///OBJECT_DATA_BACKUP_*
+}
+
+const (
+    ///Below consts defines the access mode when creating an object for write
+    OBJECT_DATA_ACCESS_MODE_UNKNOWN int = 0
+    OBJECT_DATA_ACCESS_MODE_FILE    int = 1
+    OBJECT_DATA_ACCESS_MODE_BLOCK   int = 2
+
+	OBJECT_DATA_BACKUP_MODE_UNKNOWN int = 0
+	OBJECT_DATA_BACKUP_MODE_FULL    int = 1
+	OBJECT_DATA_BACKUP_MODE_INC     int = 2    
+)
+
+///ManifestEntryMetadata is the metadata describing one manifest data
+type ManifestEntryMetadata struct {
+    ID      ID                ///The ID of the manifest data
+    Length  int               ///The data size of the manifest data
+    Labels  map[string]string ///Labels saved together with the manifest data
+    ModTime time.Time         ///Modified time of the manifest data
+}
+ 
+type RepoManifest struct {
+    Payload  interface{}            ///The user data of manifest
+    Metadata *ManifestEntryMetadata ///The metadata data of manifest
+}
+ 
+type ManifestFilter struct {
+    Labels map[string]string
+}
+```  
+
+## Workflow
+
+### Backup & Restore Workflow
+
+We preserve the bone of the existing BR workflow, that is:
+
+- Still use the Velero Server pod and PodVolumeBackup daemonSet (originally called Restic daemonset) pods to hold the corresponding controllers and modules
```

Comment:
I'm thinking we probably want a different name for the PodVolumeBackup DaemonSet, since it handles both PodVolumeBackups and PodVolumeRestore. Since the main application deployment is just called "velero" -- what about "velero-volume"?

Reply:
@Lyndon-Li Won't the DaemonSet also be used for PodVolumeRestore? SInce the deployment is called "velero", I'm thinking "velero-volume" or maybe "velero-volumes" for the daemonset.

So for the PodVolumeBackup/PodVolumeRestore CRs, I think those are fine as-is -- the key here is that with both restic and kopia, you need both a pod *and* a volume. Both are necessary for the backup or restore. If we're naming separate actions for backup and restore, PodVolumeBackup/PodVolumeRestore should be fine. If you need a more generic name to describe both actions, then PodVolume should be fine.

Also, a minor point -- the PodVolumeBackup/Restore doesn't back up/restore the pod along with volumes -- it's just for volumes. Pod is restored by the normal backup/restore process. The reason for "Pod" in the CRD name is that the CR has to reference both the pod *and* the volume, so that the right daemonset pod will pick it up (the daemonset pod on the same node as the pod named in the PR is the one that has to do the filesystem copy).

Reply:
@sseago Thank you for the suggestions. Some comments from me:

- For PodVolumeBackup/PodVolumeRestore related things, I agree with you that we keep the name as is and in the scenarios that we want a generic word, we could use PodVolume.
- For the Daemonset's name, yes, currently, it is for PodVolume backup and restore, so it is related to volumes. On the other hand, I am just thinking that we might add other functionalities that are not related to volumes and volume backup/restore in future, so what about we make it more generic, like Velero-Daemon? "Daemon" is not an unfamiliar word in either kubernetes or traditional world (i.e. the OS), it means the same all the time -- existing across the system's entire space and entire lifecycle, and do something to serve users' requests or internal requests.

For the statement "the PodVolumeBackup/Restore doesn't back up/restore the pod along with volumes", maybe "associated with" is a correct word against "along with", and yes, we are on the same page.

Reply:
We slightly prefer the name `velero-node-agent`, but it may cause confusion in TMC, let's double check with them.

Reply:
Checked with TMC offline, they are OK using the name `velero-node-agent` but have some concerns about the upgrade process.  They also suggested that we add label to the ds/pods to imply the usage and mode for the `node-agent` to avoid confusion.

Reply:
Done changing the daemonset name.
The upgrade process and labels have been added in the Upgrade section.